File: BacktrackingAlgorithm/sudokuSolver.cpp
#include <algorithm>
#include <boardtype.cpp>

class SudokuSolver
{
public:
    bool finished;
    bool possibleBuf[DIMENSION+1];

    void backtrack(int a[], int k, BoardType *board)
    {
        int c[DIMENSION];
        for (int j = 0; j < DIMENSION; j++)
        {
            c[j] = 0;
        }
        int ncandidates;
        if (board->freecount == 0)
        {
            finished = true;
        }
        else
        {
            k++;
            construct_candidates(k, board, c, &ncandidates);
            for(int i=0; i<ncandidates; i++)
            {
                a[k] = c[i];
                make_move(a, k, board);
                backtrack(a, k, board);
                if (finished) return;
                unmake_move(k, board);
            }
        }
    }

    void construct_candidates(int k, BoardType *board, int c[], int *ncandidates)
    {
        int x,y;
        bool possible[DIMENSION+1];
        for (int j = 0; j < DIMENSION+1; j++)
        {
            possible[j] = true;
        }
        next_square(&x, &y, board);
        board->move[k].x = x;
        board->move[k].y = y;
        *ncandidates = 0;
        if (x == -1) return;
        possible_values(x, y, board, possible);
        for (int i=1; i<=DIMENSION; i++)
        {
            if(possible[i] == true)
            {
                c[*ncandidates] = i;
                *ncandidates = *ncandidates + 1;
            }
        }
    }

    void make_move(int a[], int k, BoardType *board)
    {
        board->m[board->move[k].x][board->move[k].y] = a[k];
        board->freecount--;
    }

    void unmake_move(int k, BoardType *board)
    {
        board->m[board->move[k].x][board->move[k].y] = 0;
        board->freecount++;
    }

    void next_square(int* x, int* y, BoardType *board)
    {
        for(int i = 0; i < DIMENSION; i++)
        {
            for(int j = 0; j < DIMENSION; j++)
            {
                board->move[i * DIMENSION + j + 1].nCandidates = 0;
                if (board->m[i][j] == 0)
                {
                    for (int k = 0; k < DIMENSION+1; k++)
                    {
                        possibleBuf[k] = true;
                    }
                    int nCandidates = possible_values(i, j, board, possibleBuf);
                    if (nCandidates == 1)
                    {
                        *x = i;
                        *y = j;
                        return;
                    }
                    else
                    {
                        board->move[i * DIMENSION + j + 1].nCandidates = nCandidates;
                    }
                }
            }
        }

        for(int j = 2; j < DIMENSION+1; j++)
        {
            for(int i = 1; i < NCELLS+1; i++)
            {
                if(board->move[i].nCandidates == j)
                {
                    *x = (i-1)/DIMENSION;
                    *y = i-1 - *x*DIMENSION;
                    return;
                }
            }
        }

        *x = -1;
    }

    int possible_values(int x, int y, BoardType *board, bool *possible)
    {
        int i;
        for (i = 0; i < DIMENSION; i++)
        {
            if (board->m[x][i] != 0)
            {
                possible[board->m[x][i]] = false;
            }
        }
        for (i = 0; i < DIMENSION; i++)
        {
            if (board->m[i][y] != 0)
            {
                possible[board->m[i][y]] = false;
            }
        }
        int xOffset = (x/3)*3;
        int yOffset = (y/3)*3;
        for (i = xOffset; i < xOffset + 3; i++)
        {
            for (int j = yOffset; j < yOffset + 3; j++)
            {
                if (board->m[i][j] != 0)
                {
                    possible[board->m[i][j]] = false;
                }
            }
        }
        int nCandidates = 0;
        for (i = 1; i < DIMENSION+1; i++)
        {
            if (possible[i])
            {
                nCandidates++;
            }
        }
        return nCandidates;
    }
};

File: JS version/js/sudokusolver.js
import {BoardType} from '/js/boardtype.js'

export class SudokuSolver {
    constructor() {
      this.finished = false;
      this.possibleBuf = Array(BoardType.Dimension + 1).fill(false);
    }
  
    backtrack(a, k, board) {
      const c = Array(BoardType.Dimension).fill(0);
      let ncandidates;

      if (board.freecount === 0) {
        
        this.finished = true;
      } else {
        
        k++;
        ncandidates = this.constructCandidates(k, board, c);

         for (let i = 0; i < ncandidates; i++) {
            a[k] = c[i];
            this.makeMove(a, k, board);
            this.backtrack(a, k, board);
            if (this.finished) return;
            this.unmakeMove(k, board);
          }
        };
      }
  
    constructCandidates(k, board, c) {
      const possible = Array(BoardType.Dimension + 1).fill(true);
      let { x, y } = this.nextSquare(board);
      board.move[k].x = x;
      board.move[k].y = y;
  
      let ncandidates = 0;
      if (x === -1) return 0;
  
      this.possibleValues(x, y, board, possible);
      for (let i = 1; i <= BoardType.Dimension; i++) {
        if (possible[i]) {
          c[ncandidates] = i;
          ncandidates++;
        }
      }

      return ncandidates;
    }
  
    makeMove(a, k, board) {
      board.m[board.move[k].x][board.move[k].y] = a[k];
      board.freecount--;
    }
  
    unmakeMove(k, board) {
      board.m[board.move[k].x][board.move[k].y] = 0;
      board.freecount++;
    }
  
    nextSquare(board) {
      let x = -1, y = -1; 
      for (let i = 0; i < BoardType.Dimension; i++) {
        for (let j = 0; j < BoardType.Dimension; j++) {
          board.move[i * BoardType.Dimension + j + 1].nCandidates = 0;
          if (board.m[i][j] === 0) {
            const possible = Array(BoardType.Dimension + 1).fill(true);
            const nCandidates = this.possibleValues(i, j, board, possible);
            if (nCandidates === 1) {
              x = i;
              y = j;
              return { x, y }; 
            } else {
              board.move[i * BoardType.Dimension + j + 1].nCandidates = nCandidates;
            }
          }
        }
      }
  
      for (let j = 2; j < BoardType.Dimension + 1; j++) {
        for (let i = 1; i < BoardType.NCELLS + 1; i++) {
          if (board.move[i].nCandidates === j) {
            x = Math.floor((i - 1) / BoardType.Dimension);
            y = i - 1 - x * BoardType.Dimension;
            return { x, y };
          }
        }
      }
  
      return { x: -1, y: -1 };
    }
  
    possibleValues(x, y, board, possible) {  
      for (let i = 0; i < BoardType.Dimension; i++) {
        if (board.m[x][i] !== 0) {
          possible[board.m[x][i]] = false;
        }
      }
  
      for (let i = 0; i < BoardType.Dimension; i++) {
        if (board.m[i][y] !== 0) {
          possible[board.m[i][y]] = false;
        }
      }
  
      const xOffset = Math.floor(x / 3) * 3;
      const yOffset = Math.floor(y / 3) * 3;
      for (let i = xOffset; i < xOffset + 3; i++) {
        for (let j = yOffset; j < yOffset + 3; j++) {
          if (board.m[i][j] !== 0) {
            possible[board.m[i][j]] = false;
          }
        }
      }
  
      let nCandidates = 0;
      for (let i = 1; i < BoardType.Dimension + 1; i++) {
        if (possible[i]) {
          nCandidates++;
        }
      }
  
      return nCandidates;
    }
  }
  